mat64/list_test.go

```diff
@@ -0,0 +1,706 @@
+				Data: make([]float64, len(t.mat.Data)),
+			},
+			n: t.n,
+		}
+		copy(m.mat.Data, t.mat.Data)
+		return m
+	case Transpose:
+		matrix := t.Matrix
+		mat := matrix.(*Vector) // Should only be here for the vector type
+		m := &Vector{
+			mat: blas64.Vector{
+				Inc:  mat.mat.Inc,
+				Data: make([]float64, len(mat.mat.Data)),
+			},
+			n: mat.n,
+		}
+		copy(m.mat.Data, mat.mat.Data)
+		return Transpose{m}
+	}
+}
+
+// makeTransposeCopyOf makes a transpose copy of the given matrix with the same
+// underlying type.
+func makeTransposeCopyOf(a Matrix) Matrix {
+	switch t := a.(type) {
+	default:
+		panic("unknown type in make transpose of")
+	case *Dense, *basicMatrix, *basicVectorer:
+		r, c := a.Dims()
+		m := NewDense(c, r, nil)
+		for i := 0; i < r; i++ {
+			for j := 0; j < c; j++ {
+				m.Set(j, i, a.At(i, j))
+			}
+		}
+		return returnAs(m, t)
+	case *SymDense, *basicSymmetric:
+		return makeCopyOf(a)
+	case *TriDense, *basicTriangular:
+		n, upper := t.(Triangular).Triangle()
+		mUpper := !upper
+		m := NewTriDense(n, mUpper, nil)
+		if mUpper {
+			for i := 0; i < n; i++ {
+				for j := i; j < n; j++ {
+					m.SetTri(i, j, t.At(j, i))
+				}
+			}
+		} else {
+			for i := 0; i < n; i++ {
+				for j := 0; j <= i; j++ {
+					m.SetTri(i, j, t.At(j, i))
+				}
+			}
+		}
+		return returnAs(m, t)
+	case *Vector:
+		m := makeCopyOf(a)
+		return Transpose{m}
+	case Transpose:
+		v := t.Matrix.(*Vector)
+		m := makeCopyOf(v)
+		return m
+	}
+}
+
+// sameType returns true if a and b have the same underlying type.
+func sameType(a, b Matrix) bool {
+	if reflect.ValueOf(a).Type() == reflect.ValueOf(b).Type() {
+		return true
+	}
+	return false
+}
+
+// canShadow returns true if the two matrices could be represented by the same
+// pointer.
+func canShadow(receiver, a Matrix, trans bool) bool {
+	rr, rc := receiver.Dims()
+	ar, ac := a.Dims()
+	if !sameType(receiver, a) {
+		return false
+	}
+	if !trans {
+		if rr == ar && rc == ac {
+			return true
+		}
+		return false
+	}
+	if rr == ac && rc == ar {
+		return true
+	}
+	return false
+}
+
+// equalApprox returns whether the elements of a and b are the same to within
+// the tolerance.
+func equalApprox(a, b Matrix, tol float64) bool {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br {
+		return false
+	}
+	if ac != bc {
+		return false
+	}
+	for i := 0; i < ar; i++ {
+		for j := 0; j < ac; j++ {
+			if !floats.EqualWithinAbsOrRel(a.At(i, j), b.At(i, j), tol, tol) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// equal returns true if the matrices have equal entries.
+func equal(a, b Matrix) bool {
+	ar, ac := a.Dims()
+	br, bc := b.Dims()
+	if ar != br {
+		return false
+	}
+	if ac != bc {
+		return false
+	}
+	for i := 0; i < ar; i++ {
+		for j := 0; j < ac; j++ {
+			if a.At(i, j) != b.At(i, j) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+// testTwoInput tests a method that has two input arguments.
+// receiver is a value of the receiver type.
```

Comment:
The docs here are confusing; there is a claim that method returns false for some condition, but this is clearly not true.

I'm going to suggest something that I've never seen before, but I think will work here: make parameter doc comments look like a public struct definition, e.g.

```
// testTwoInput tests a method that has two input arguments.
func testTwoInput(c *check.C,
	// receiver is a value of the receiver type.
	receiver Matrix,

	// method is the generalied receiver.Method(a,b).
	method func(receiver, a, b Matrix),

	// denseComparison performs the same operation
	// as method, but with dense matrices for
	// comparison with the result.
	denseComparison func(receiver, a, b *Dense),

	// legalTypes returns whether the concrete types
	// in Matrix are valid for the method.
	legalTypes func(a, b Matrix) bool,

	// legalSize returns whether the matrix sizes are
	// valid for the method.
	legalSize func(ar, ac, br, bc int) bool,
) {
    ...
```